File: exercism-python/luhn/luhn.py
import re


class Luhn(object):
    """
    The Luhn algorithm is a simple checksum formula used to validate
    a variety of identification numbers, such as credit card numbers
    and Canadian Social Insurance Numbers.
    """
    SPACE = ' '

    def __init__(self, card_num):
        self.card_num = card_num

    def is_valid(self):
        """
        Determine whether or not the number is valid per the Luhn formula.
        """

        card_num = self.card_num.replace(self.SPACE, '')
        match = re.search(r'^\d{2,}$', card_num)
        if not match:
            return False
        digits = [int(n) for n in match.group()]

        sum = 0
        alternate = False
        for d in reversed(digits):
            if alternate:
                d = d * 2
                if d > 9:
                    d -= 9
            sum += d
            alternate = not alternate

        return sum % 10 == 0

File: exercism-go/luhn/luhn.go
package luhn

import (
	"regexp"
	"strconv"
	"strings"
)

// Valid determines whether or not a number is valid per the Luhn formula.
func Valid(id string) bool {
	id = strings.Replace(id, " ", "", -1)
	re := regexp.MustCompile(`^\d{2,}$`)
	if !re.MatchString(id) {
		return false
	}

	arr := strings.Split(id, "")
	sum := 0
	alternate := false

	for i := len(arr) - 1; i >= 0; i-- {
		n, _ := strconv.Atoi(arr[i])
		// if err != nil {
		// 	return false
		// }
		if alternate {
			n <<= 1
			if n > 9 {
				n -= 9
			}
		}
		alternate = !alternate
		sum += n
	}

	return (sum%10 == 0)
}
